File: weather-app/src/main/java/rcgonzalezf/org/weather/common/analytics/observer/LogcatAnalyticsObserver.java
package rcgonzalezf.org.weather.common.analytics.observer;

import android.support.annotation.NonNull;
import android.util.Log;
import java.util.Map;
import rcgonzalezf.org.weather.common.analytics.AnalyticsBaseData;
import rcgonzalezf.org.weather.common.analytics.AnalyticsEvent;
import rcgonzalezf.org.weather.common.analytics.AnalyticsObserver;

public class LogcatAnalyticsObserver implements AnalyticsObserver {

  private static String TAG = LogcatAnalyticsObserver.class.getName();

  @Override
  public void onScreen(@NonNull String screenName, @NonNull AnalyticsBaseData analyticsBaseData) {
    String builder = "onScreen: [ " +
        screenName +
        " | data: " +
        printableData(analyticsBaseData.data()) +
        " ]";

    Log.d(TAG, builder);
  }

  @Override public void onAction(@NonNull AnalyticsEvent analyticsEvent, @NonNull String screenName,
      @NonNull AnalyticsBaseData analyticsBaseData) {
    String builder = "onAction: [ event:" +
        " name=" + analyticsEvent.name +
        " additionalValue=" + analyticsEvent.additionalValue +
        " | screenName, " +
        screenName +
        " | data: " +
        printableData(analyticsBaseData.data()) +
        " ]";
    Log.d(TAG, builder);
  }

  private StringBuilder printableData(Map<String, String> data) {
    StringBuilder stringBuilder = new StringBuilder();
    for (String key : data.keySet()) {
      stringBuilder.append("{").append(key).append(" | ").append(data.get(key)).append("}");
    }
    return stringBuilder;
  }
}

File: weather-app/src/main/java/rcgonzalezf/org/weather/common/analytics/observer/LogcatAnalyticsObserver.kt
package rcgonzalezf.org.weather.common.analytics.observer

import android.util.Log
import rcgonzalezf.org.weather.common.analytics.AnalyticsBaseData
import rcgonzalezf.org.weather.common.analytics.AnalyticsEvent
import rcgonzalezf.org.weather.common.analytics.AnalyticsObserver

class LogcatAnalyticsObserver : AnalyticsObserver {

    companion object {
        private val TAG = LogcatAnalyticsObserver::class.java.name
    }

    override fun onScreen(screenName: String, analyticsBaseData: AnalyticsBaseData) {
        val builder = "onScreen: [ " +
                screenName +
                " | data: " +
                printableData(analyticsBaseData.data()) +
                " ]"
        Log.d(TAG, builder)
    }

    override fun onAction(analyticsEvent: AnalyticsEvent, screenName: String,
                          analyticsBaseData: AnalyticsBaseData) {
        val builder = "onAction: [ event:" +
                " name=" + analyticsEvent.name +
                " additionalValue=" + analyticsEvent.additionalValue +
                " | screenName, " +
                screenName +
                " | data: " +
                printableData(analyticsBaseData.data()) +
                " ]"
        Log.d(TAG, builder)
    }

    private fun printableData(data: Map<String, String>): StringBuilder {
        val stringBuilder = StringBuilder()
        for (key in data.keys) {
            stringBuilder.append("{").append(key).append(" | ").append(data[key]).append("}")
        }
        return stringBuilder
    }
}
